File: scripts/ingest.mjs
/**
 * Austin VC & Startup Map — EXA Ingest Script
 *
 * Queries the EXA API for Austin-based startups and VC firms,
 * then upserts the results into Supabase.
 *
 * Usage:
 *   1. Copy .env.example → .env and fill in your keys
 *   2. npm install
 *   3. npm run ingest
 */

import "dotenv/config";
import Exa from "exa-js";
import { createClient } from "@supabase/supabase-js";

// ── Config ─────────────────────────────────────────────────────
const exa = new Exa(process.env.EXA_API_KEY);

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_SERVICE_KEY // service role for RLS bypass
);

// ── EXA Queries ────────────────────────────────────────────────
const QUERIES = [
  // Startups
  { query: "Austin Texas startups founded recently", type: "startup" },
  { query: "Austin TX seed stage startup companies", type: "startup" },
  { query: "early stage startups based in Austin Texas", type: "startup" },
  { query: "Austin tech startup companies 2024 2025", type: "startup" },
  { query: "Austin Texas B2B SaaS startups", type: "startup" },
  { query: "Austin fintech healthtech startups", type: "startup" },

  // VCs
  { query: "venture capital firms in Austin Texas", type: "vc" },
  { query: "Austin TX VC investors and venture funds", type: "vc" },
  { query: "Austin angel investors and seed funds", type: "vc" },
  { query: "Austin Texas venture capital partners", type: "vc" },
];

// ── Helpers ────────────────────────────────────────────────────

/**
 * Normalize a URL to its bare domain for deduplication.
 * "https://www.example.com/about" → "example.com"
 */
function domainOf(url) {
  try {
    const host = new URL(url).hostname.replace(/^www\./, "");
    return host;
  } catch {
    return url;
  }
}

/**
 * Attempt to extract a lat/lng from the EXA result text.
 * Falls back to central Austin coordinates.
 *
 * In production you'd call a geocoding API (Google Maps, Mapbox, etc.)
 * using the extracted address. For now we jitter around downtown Austin
 * so every pin is visible on the map.
 */
function extractLocation(result) {
  // Default: downtown Austin with small random jitter so pins don't stack
  const baseLat = 30.2672;
  const baseLng = -97.7431;
  const jitter = () => (Math.random() - 0.5) * 0.06; // ~3 km spread

  return {
    lat: baseLat + jitter(),
    lng: baseLng + jitter(),
  };
}

/**
 * Parse relevant fields from an EXA result + its summary/text.
 */
function parseResult(result, queryType) {
  const { lat, lng } = extractLocation(result);

  return {
    name: result.title || "Unknown",
    type: queryType,
    website: result.url || null,
    description: result.summary || result.text?.slice(0, 500) || null,
    exa_source_url: result.url || null,
    location: `SRID=4326;POINT(${lng} ${lat})`, // WKT for PostGIS
    city: "Austin",
    state: "TX",
  };
}

// ── Main Ingest Logic ──────────────────────────────────────────

async function runQuery({ query, type }) {
  console.log(`\n🔍  Querying EXA: "${query}" [${type}]`);

  const response = await exa.searchAndContents(query, {
    category: "company",
    numResults: 30,
    contents: {
      text: { maxCharacters: 2000 },
      summary: true,
    },
  });

  console.log(`   ↳ Got ${response.results.length} results`);
  return response.results.map((r) => ({ ...r, _queryType: type, _queryUsed: query }));
}

async function upsertOrganization(org) {
  // Upsert by website (dedupe key)
  const { data, error } = await supabase
    .from("organizations")
    .upsert(org, { onConflict: "website", ignoreDuplicates: false })
    .select("id")
    .single();

  if (error) {
    // Duplicate or constraint error — try to fetch existing
    if (error.code === "23505" || error.message?.includes("duplicate")) {
      const { data: existing } = await supabase
        .from("organizations")
        .select("id")
        .eq("website", org.website)
        .single();
      return existing?.id;
    }
    console.warn(`   ⚠️  Upsert failed for "${org.name}": ${error.message}`);
    return null;
  }

  return data?.id;
}

async function saveExaResult(result, organizationId) {
  await supabase.from("exa_search_results").insert({
    organization_id: organizationId,
    query_used: result._queryUsed,
    score: result.score ?? null,
    url: result.url,
    title: result.title,
    published_date: result.publishedDate || null,
    raw_result: result,
  });
}

async function ingest() {
  console.log("🚀  Austin VC & Startup Map — EXA Ingest");
  console.log("=========================================\n");

  const seen = new Set(); // track domains we've already processed
  let totalInserted = 0;
  let totalSkipped = 0;

  for (const q of QUERIES) {
    let results;
    try {
      results = await runQuery(q);
    } catch (err) {
      console.error(`   ❌  EXA query failed: ${err.message}`);
      continue;
    }

    for (const result of results) {
      const domain = domainOf(result.url);

      // Skip duplicates within this run
      if (seen.has(domain)) {
        totalSkipped++;
        continue;
      }
      seen.add(domain);

      const org = parseResult(result, q.type);
      if (!org.website) continue;

      const orgId = await upsertOrganization(org);
      if (orgId) {
        await saveExaResult(result, orgId);
        totalInserted++;
        console.log(`   ✅  ${org.name} (${org.type})`);
      }
    }

    // Rate-limit: small delay between queries
    await new Promise((r) => setTimeout(r, 500));
  }

  console.log(`\n=========================================`);
  console.log(`✅  Done! Inserted/updated: ${totalInserted} | Skipped dupes: ${totalSkipped}`);
}

ingest().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: scripts/ingest.py
"""
Austin VC & Startup Map — EXA Ingest Script

Queries the EXA API for Austin-based startups and VC firms,
then upserts the results into Supabase.

Usage:
    1. Copy .env.example → .env and fill in your keys
    2. pip install -r requirements.txt
    3. python scripts/ingest.py
"""

import os
import random
import json
from datetime import datetime

from dotenv import load_dotenv
from exa_py import Exa
from supabase import create_client

load_dotenv()

# ── Config ─────────────────────────────────────────────────────
exa = Exa(api_key=os.environ.get("EXA_API_KEY"))

supabase = create_client(
    os.environ.get("SUPABASE_URL"),
    os.environ.get("SUPABASE_SERVICE_KEY"),  # service role for RLS bypass
)

# ── EXA Queries ────────────────────────────────────────────────
QUERIES = [
    # Startups
    {"query": "Austin Texas startup company", "type": "startup"},
    {"query": "Austin TX seed stage startup", "type": "startup"},
    {"query": "early stage startup based in Austin Texas", "type": "startup"},
    {"query": "Austin tech startup 2024 2025", "type": "startup"},
    {"query": "Austin Texas B2B SaaS startup", "type": "startup"},
    {"query": "Austin fintech healthtech startup", "type": "startup"},
    # VCs
    {"query": "venture capital firm in Austin Texas", "type": "vc"},
    {"query": "Austin TX VC investor venture fund", "type": "vc"},
    {"query": "Austin angel investor seed fund", "type": "vc"},
    {"query": "Austin Texas venture capital partner", "type": "vc"},
]


# ── Helpers ────────────────────────────────────────────────────

def domain_of(url: str) -> str:
    """Normalize a URL to its bare domain for deduplication."""
    try:
        from urllib.parse import urlparse
        host = urlparse(url).hostname or url
        if host.startswith("www."):
            host = host[4:]
        return host
    except Exception:
        return url


def extract_location():
    """
    Returns jittered downtown Austin coordinates.
    In production, replace this with a geocoding API call
    (Google Maps, Mapbox, etc.) using the company address.
    """
    base_lat = 30.2672
    base_lng = -97.7431
    jitter = lambda: (random.random() - 0.5) * 0.06  # ~3 km spread
    return base_lat + jitter(), base_lng + jitter()


def parse_result(result, query_type: str) -> dict:
    """Parse relevant fields from an EXA result."""
    lat, lng = extract_location()

    # Extract description from highlights or text
    description = None
    if hasattr(result, "highlights") and result.highlights:
        description = " ".join(result.highlights)[:500]
    elif hasattr(result, "text") and result.text:
        description = result.text[:500]

    return {
        "name": result.title or "Unknown",
        "type": query_type,
        "website": result.url or None,
        "description": description,
        "exa_source_url": result.url or None,
        "location": f"SRID=4326;POINT({lng} {lat})",  # WKT for PostGIS
        "city": "Austin",
        "state": "TX",
    }


# ── Main Ingest Logic ──────────────────────────────────────────

def run_query(query_config: dict) -> list:
    """Run a single EXA search query and return results."""
    query = query_config["query"]
    qtype = query_config["type"]
    print(f"\n🔍  Querying EXA: \"{query}\" [{qtype}]")

    results = exa.search_and_contents(
        query,
        type="auto",
        category="company",
        num_results=30,
        highlights={"max_characters": 4000},
    )

    print(f"   ↳ Got {len(results.results)} results")

    # Tag each result with metadata
    tagged = []
    for r in results.results:
        r._query_type = qtype
        r._query_used = query
        tagged.append(r)
    return tagged


def upsert_organization(org: dict) -> str | None:
    """Upsert an organization by website (dedupe key). Returns the org ID."""
    try:
        response = (
            supabase.table("organizations")
            .upsert(org, on_conflict="website")
            .execute()
        )
        if response.data:
            return response.data[0].get("id")
    except Exception as e:
        if "duplicate" in str(e).lower() or "23505" in str(e):
            # Already exists — fetch the existing ID
            existing = (
                supabase.table("organizations")
                .select("id")
                .eq("website", org["website"])
                .single()
                .execute()
            )
            if existing.data:
                return existing.data.get("id")
        print(f"   ⚠️  Upsert failed for \"{org['name']}\": {e}")
    return None


def save_exa_result(result, organization_id: str):
    """Save the raw EXA result for audit/reprocessing."""
    raw = {
        "title": result.title,
        "url": result.url,
        "score": result.score if hasattr(result, "score") else None,
        "published_date": result.published_date if hasattr(result, "published_date") else None,
        "highlights": result.highlights if hasattr(result, "highlights") else None,
    }

    supabase.table("exa_search_results").insert({
        "organization_id": organization_id,
        "query_used": result._query_used,
        "score": result.score if hasattr(result, "score") else None,
        "url": result.url,
        "title": result.title,
        "published_date": result.published_date if hasattr(result, "published_date") else None,
        "raw_result": raw,
    }).execute()


def ingest():
    """Main entry point — run all queries and upsert results."""
    print("🚀  Austin VC & Startup Map — EXA Ingest")
    print("=" * 45, "\n")

    seen: set[str] = set()  # track domains already processed
    total_inserted = 0
    total_skipped = 0

    for q in QUERIES:
        try:
            results = run_query(q)
        except Exception as err:
            print(f"   ❌  EXA query failed: {err}")
            continue

        for result in results:
            domain = domain_of(result.url)

            # Skip duplicates within this run
            if domain in seen:
                total_skipped += 1
                continue
            seen.add(domain)

            org = parse_result(result, q["type"])
            if not org["website"]:
                continue

            org_id = upsert_organization(org)
            if org_id:
                save_exa_result(result, org_id)
                total_inserted += 1
                print(f"   ✅  {org['name']} ({org['type']})")

    print(f"\n{'=' * 45}")
    print(f"✅  Done! Inserted/updated: {total_inserted} | Skipped dupes: {total_skipped}")


if __name__ == "__main__":
    ingest()
